File: artifacts/heman/decompose/src/plane.c
kmPlane *kmPlaneFill(kmPlane *plane, float a, float b, float c, float d)
{
  plane->a = a;
  plane->b = b;
  plane->c = c;
  plane->d = d;
  return plane;
}


/*** DEPENDENCIES:
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


kmPlane *kmPlaneScale(kmPlane *pOut, const kmPlane *pP, float s)
{
  assert(0 && "Not implemented");
  return 0;
}


/*** DEPENDENCIES:
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


float kmPlaneDot(const kmPlane *pP, const kmVec4 *pV)
{
  return (((pP->a * pV->x) + (pP->b * pV->y)) + (pP->c * pV->z)) + (pP->d * pV->w);
}


/*** DEPENDENCIES:
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
typedef struct kmVec4
{
  float x;
  float y;
  float z;
  float w;
} kmVec4
----------------------------
***/


float kmPlaneDotCoord(const kmPlane *pP, const kmVec3 *pV)
{
  return (((pP->a * pV->x) + (pP->b * pV->y)) + (pP->c * pV->z)) + pP->d;
}


/*** DEPENDENCIES:
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


float kmPlaneDotNormal(const kmPlane *pP, const kmVec3 *pV)
{
  return ((pP->a * pV->x) + (pP->b * pV->y)) + (pP->c * pV->z);
}


/*** DEPENDENCIES:
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


kmPlane *kmPlaneFromNormalAndDistance(kmPlane *plane, const struct kmVec3 *normal, const float dist)
{
  plane->a = normal->x;
  plane->b = normal->y;
  plane->c = normal->z;
  plane->d = dist;
  return plane;
}


/*** DEPENDENCIES:
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


KM_POINT_CLASSIFICATION kmPlaneClassifyPoint(const kmPlane *pIn, const kmVec3 *pP)
{
  float distance = (((pIn->a * pP->x) + (pIn->b * pP->y)) + (pIn->c * pP->z)) + pIn->d;
  if (distance > 0.0001)
  {
    return POINT_INFRONT_OF_PLANE;
  }
  if (distance < (-0.0001))
  {
    return POINT_BEHIND_PLANE;
  }
  return POINT_ON_PLANE;
}


/*** DEPENDENCIES:
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


kmPlane *kmPlaneFromPointAndNormal(kmPlane *pOut, const kmVec3 *pPoint, const kmVec3 *pNormal)
{
  pOut->a = pNormal->x;
  pOut->b = pNormal->y;
  pOut->c = pNormal->z;
  pOut->d = -kmVec3Dot(pNormal, pPoint);
  return pOut;
}


/*** DEPENDENCIES:
float kmVec3Dot(const kmVec3 *pV1, const kmVec3 *pV2)
{
  return ((pV1->x * pV2->x) + (pV1->y * pV2->y)) + (pV1->z * pV2->z);
}


----------------------------
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


kmPlane *kmPlaneNormalize(kmPlane *pOut, const kmPlane *pP)
{
  kmVec3 n;
  float l = 0;
  if (((!pP->a) && (!pP->b)) && (!pP->c))
  {
    pOut->a = pP->a;
    pOut->b = pP->b;
    pOut->c = pP->c;
    pOut->d = pP->d;
    return pOut;
  }
  n.x = pP->a;
  n.y = pP->b;
  n.z = pP->c;
  l = 1.0f / kmVec3Length(&n);
  kmVec3Normalize(&n, &n);
  pOut->a = n.x;
  pOut->b = n.y;
  pOut->c = n.z;
  pOut->d = pP->d * l;
  return pOut;
}


/*** DEPENDENCIES:
float kmVec3Length(const kmVec3 *pIn)
{
  return sqrtf((kmSQR(pIn->x) + kmSQR(pIn->y)) + kmSQR(pIn->z));
}


----------------------------
kmVec3 *kmVec3Normalize(kmVec3 *pOut, const kmVec3 *pIn)
{
  if (((!pIn->x) && (!pIn->y)) && (!pIn->z))
  {
    return kmVec3Assign(pOut, pIn);
  }
  float l = 1.0f / kmVec3Length(pIn);
  kmVec3 v;
  v.x = pIn->x * l;
  v.y = pIn->y * l;
  v.z = pIn->z * l;
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


kmPlane *kmPlaneExtractFromMat4(kmPlane *pOut, const struct kmMat4 *pIn, int row)
{
  int scale = (row < 0) ? (-1) : (1);
  row = abs(row) - 1;
  pOut->a = pIn->mat[3] + (scale * pIn->mat[row]);
  pOut->b = pIn->mat[7] + (scale * pIn->mat[row + 4]);
  pOut->c = pIn->mat[11] + (scale * pIn->mat[row + 8]);
  pOut->d = pIn->mat[15] + (scale * pIn->mat[row + 12]);
  return kmPlaneNormalize(pOut, pOut);
}


/*** DEPENDENCIES:
typedef struct kmMat4
{
  float mat[16];
} kmMat4
----------------------------
kmPlane *kmPlaneNormalize(kmPlane *pOut, const kmPlane *pP)
{
  kmVec3 n;
  float l = 0;
  if (((!pP->a) && (!pP->b)) && (!pP->c))
  {
    pOut->a = pP->a;
    pOut->b = pP->b;
    pOut->c = pP->c;
    pOut->d = pP->d;
    return pOut;
  }
  n.x = pP->a;
  n.y = pP->b;
  n.z = pP->c;
  l = 1.0f / kmVec3Length(&n);
  kmVec3Normalize(&n, &n);
  pOut->a = n.x;
  pOut->b = n.y;
  pOut->c = n.z;
  pOut->d = pP->d * l;
  return pOut;
}


----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
***/


kmVec3 *kmPlaneIntersectLine(kmVec3 *pOut, const kmPlane *pP, const kmVec3 *pV1, const kmVec3 *pV2)
{
  unsigned int pOut_idx = 0;
  kmVec3 d;
  kmVec3Subtract(&d, pV2, pV1);
  kmVec3 n;
  n.x = pP->a;
  n.y = pP->b;
  n.z = pP->c;
  kmVec3Normalize(&n, &n);
  float nt = -((((n.x * pV1->x) + (n.y * pV1->y)) + (n.z * pV1->z)) + pP->d);
  float dt = ((n.x * d.x) + (n.y * d.y)) + (n.z * d.z);
  if (fabs(dt) < 0.0001)
  {
    pOut_idx = 0;
    return pOut;
  }
  float t = nt / dt;
  pOut->x = pV1->x + (d.x * t);
  pOut->y = pV1->y + (d.y * t);
  pOut->z = pV1->z + (d.z * t);
  return pOut;
}


/*** DEPENDENCIES:
kmVec3 *kmVec3Normalize(kmVec3 *pOut, const kmVec3 *pIn)
{
  if (((!pIn->x) && (!pIn->y)) && (!pIn->z))
  {
    return kmVec3Assign(pOut, pIn);
  }
  float l = 1.0f / kmVec3Length(pIn);
  kmVec3 v;
  v.x = pIn->x * l;
  v.y = pIn->y * l;
  v.z = pIn->z * l;
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
kmVec3 *kmVec3Subtract(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2)
{
  kmVec3 v;
  v.x = pV1->x - pV2->x;
  v.y = pV1->y - pV2->y;
  v.z = pV1->z - pV2->z;
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
***/


kmPlane *kmPlaneFromPoints(kmPlane *pOut, const kmVec3 *p1, const kmVec3 *p2, const kmVec3 *p3)
{
  kmVec3 n;
  kmVec3 v1;
  kmVec3 v2;
  kmVec3Subtract(&v1, p2, p1);
  kmVec3Subtract(&v2, p3, p1);
  kmVec3Cross(&n, &v1, &v2);
  kmVec3Normalize(&n, &n);
  pOut->a = n.x;
  pOut->b = n.y;
  pOut->c = n.z;
  pOut->d = kmVec3Dot(kmVec3Scale(&n, &n, -1.0), p1);
  return pOut;
}


/*** DEPENDENCIES:
kmVec3 *kmVec3Normalize(kmVec3 *pOut, const kmVec3 *pIn)
{
  if (((!pIn->x) && (!pIn->y)) && (!pIn->z))
  {
    return kmVec3Assign(pOut, pIn);
  }
  float l = 1.0f / kmVec3Length(pIn);
  kmVec3 v;
  v.x = pIn->x * l;
  v.y = pIn->y * l;
  v.z = pIn->z * l;
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
kmVec3 *kmVec3Cross(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2)
{
  kmVec3 v;
  v.x = (pV1->y * pV2->z) - (pV1->z * pV2->y);
  v.y = (pV1->z * pV2->x) - (pV1->x * pV2->z);
  v.z = (pV1->x * pV2->y) - (pV1->y * pV2->x);
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
kmVec3 *kmVec3Scale(kmVec3 *pOut, const kmVec3 *pIn, const float s)
{
  pOut->x = pIn->x * s;
  pOut->y = pIn->y * s;
  pOut->z = pIn->z * s;
  return pOut;
}


----------------------------
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
float kmVec3Dot(const kmVec3 *pV1, const kmVec3 *pV2)
{
  return ((pV1->x * pV2->x) + (pV1->y * pV2->y)) + (pV1->z * pV2->z);
}


----------------------------
kmVec3 *kmVec3Subtract(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2)
{
  kmVec3 v;
  v.x = pV1->x - pV2->x;
  v.y = pV1->y - pV2->y;
  v.z = pV1->z - pV2->z;
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
***/


kmVec3 *kmPlaneGetIntersection(kmVec3 *pOut, const kmPlane *p1, const kmPlane *p2, const kmPlane *p3)
{
  kmVec3 n1;
  kmVec3 n2;
  kmVec3 n3;
  kmVec3 cross;
  kmVec3 r1;
  kmVec3 r2;
  kmVec3 r3;
  double denom = 0;
  kmVec3Fill(&n1, p1->a, p1->b, p1->c);
  kmVec3Fill(&n2, p2->a, p2->b, p2->c);
  kmVec3Fill(&n3, p3->a, p3->b, p3->c);
  kmVec3Cross(&cross, &n2, &n3);
  denom = kmVec3Dot(&n1, &cross);
  if (kmAlmostEqual(denom, 0.0))
  {
    return 0;
  }
  kmVec3Cross(&r1, &n2, &n3);
  kmVec3Cross(&r2, &n3, &n1);
  kmVec3Cross(&r3, &n1, &n2);
  kmVec3Scale(&r1, &r1, -p1->d);
  kmVec3Scale(&r2, &r2, p2->d);
  kmVec3Scale(&r3, &r3, p3->d);
  kmVec3Subtract(pOut, &r1, &r2);
  kmVec3Subtract(pOut, pOut, &r3);
  kmVec3Scale(pOut, pOut, 1.0 / denom);
  return pOut;
}


/*** DEPENDENCIES:
typedef struct kmPlane
{
  float a;
  float b;
  float c;
  float d;
} kmPlane
----------------------------
kmVec3 *kmVec3Cross(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2)
{
  kmVec3 v;
  v.x = (pV1->y * pV2->z) - (pV1->z * pV2->y);
  v.y = (pV1->z * pV2->x) - (pV1->x * pV2->z);
  v.z = (pV1->x * pV2->y) - (pV1->y * pV2->x);
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
kmVec3 *kmVec3Scale(kmVec3 *pOut, const kmVec3 *pIn, const float s)
{
  pOut->x = pIn->x * s;
  pOut->y = pIn->y * s;
  pOut->z = pIn->z * s;
  return pOut;
}


----------------------------
typedef struct kmVec3
{
  float x;
  float y;
  float z;
} kmVec3
----------------------------
kmVec3 *kmVec3Fill(kmVec3 *pOut, float x, float y, float z)
{
  pOut->x = x;
  pOut->y = y;
  pOut->z = z;
  return pOut;
}


----------------------------
unsigned char kmAlmostEqual(float lhs, float rhs)
{
  return ((lhs + 0.0001) > rhs) && ((lhs - 0.0001) < rhs);
}


----------------------------
float kmVec3Dot(const kmVec3 *pV1, const kmVec3 *pV2)
{
  return ((pV1->x * pV2->x) + (pV1->y * pV2->y)) + (pV1->z * pV2->z);
}


----------------------------
kmVec3 *kmVec3Subtract(kmVec3 *pOut, const kmVec3 *pV1, const kmVec3 *pV2)
{
  kmVec3 v;
  v.x = pV1->x - pV2->x;
  v.y = pV1->y - pV2->y;
  v.z = pV1->z - pV2->z;
  pOut->x = v.x;
  pOut->y = v.y;
  pOut->z = v.z;
  return pOut;
}


----------------------------
***/

File: artifacts/heman/translate/src/plane.rs
use crate::*;
use crate::KmPlane;
use crate::KmPointClassification;
use crate::KmVec3;
use std::f32;
use std::cmp::Ordering;

pub fn kmPlaneFill(plane: &mut KmPlane, a: f32, b: f32, c: f32, d: f32) -> &mut KmPlane {
    plane.a = a;
    plane.b = b;
    plane.c = c;
    plane.d = d;
    plane
}
pub fn kmPlaneScale<'a>(pOut: Option<&'a mut KmPlane>, pP: Option<&'a KmPlane>, s: f32) -> Option<&'a mut KmPlane> {
    // Use assert_eq! to match C's assert behavior (panic if condition fails)
    assert!(false, "Not implemented");
    Option::None
}
pub fn kmPlaneDot(pP: Option<&KmPlane>, pV: Option<&KmVec4>) -> Option<f32> {
    // Check for NULL pointers (None in Rust)
    if pP.is_none() || pV.is_none() {
        return None;
    }

    // Unwrap safely since we've checked for None
    let pP = pP.unwrap();
    let pV = pV.unwrap();

    // Perform the calculation as in C
    Some((pP.a * pV.x) + (pP.b * pV.y) + (pP.c * pV.z) + (pP.d * pV.w))
}
pub fn kmPlaneDotCoord(pP: Option<&KmPlane>, pV: Option<&KmVec3>) -> Option<f32> {
    // Check for NULL pointers (None in Rust)
    let pP = pP?;
    let pV = pV?;

    // Perform the calculation as in C
    Some((pP.a * pV.x) + (pP.b * pV.y) + (pP.c * pV.z) + pP.d)
}
pub fn kmPlaneDotNormal(pP: Option<&KmPlane>, pV: Option<&KmVec3>) -> Option<f32> {
    // Check for NULL pointers (None in Rust)
    if pP.is_none() || pV.is_none() {
        return None;
    }

    // Unwrap safely since we've checked for None
    let pP = pP.unwrap();
    let pV = pV.unwrap();

    // Perform the dot product calculation
    Some((pP.a * pV.x) + (pP.b * pV.y) + (pP.c * pV.z))
}
pub fn kmPlaneFromNormalAndDistance<'a>(
    plane: &'a mut KmPlane,
    normal: Option<&'a KmVec3>,
    dist: f32,
) -> Option<&'a mut KmPlane> {
    // Check for NULL equivalent (None in Rust)
    let normal = normal?;

    // Assign values to the plane
    plane.a = normal.x;
    plane.b = normal.y;
    plane.c = normal.z;
    plane.d = dist;

    Some(plane)
}
pub fn kmPlaneFromPointAndNormal<'a>(
    pOut: Option<&'a mut KmPlane>,
    pPoint: Option<&'a KmVec3>,
    pNormal: Option<&'a KmVec3>,
) -> Option<&'a mut KmPlane> {
    // Check for None (equivalent to NULL checks in C)
    let pOut = pOut?;
    let pPoint = pPoint?;
    let pNormal = pNormal?;

    pOut.a = pNormal.x;
    pOut.b = pNormal.y;
    pOut.c = pNormal.z;
    pOut.d = -kmVec3Dot(Some(pNormal), Some(pPoint))?;

    Some(pOut)
}

pub fn kmPlaneClassifyPoint(pIn: Option<&KmPlane>, pP: Option<&KmVec3>) -> KmPointClassification {
    // Check for NULL pointers (None in Rust)
    if pIn.is_none() || pP.is_none() {
        return KmPointClassification::POINT_ON_PLANE; // Default behavior when inputs are invalid
    }

    // Unwrap safely since we've checked for None
    let pIn = pIn.unwrap();
    let pP = pP.unwrap();

    // Calculate distance using the plane equation
    let distance = ((pIn.a * pP.x) + (pIn.b * pP.y) + (pIn.c * pP.z)) + pIn.d;

    // Classify the point based on distance
    if distance > 0.0001 {
        KmPointClassification::POINT_INFRONT_OF_PLANE
    } else if distance < -0.0001 {
        KmPointClassification::POINT_BEHIND_PLANE
    } else {
        KmPointClassification::POINT_ON_PLANE
    }
}
pub fn kmPlaneNormalize<'a>(pOut: Option<&'a mut KmPlane>, pP: Option<&'a KmPlane>) -> Option<&'a mut KmPlane> {
    // Check if either input is None (equivalent to NULL check in C)
    let pOut = pOut?;
    let pP = pP?;

    // Check if all components are zero (lines 5-12 in C)
    if (pP.a == 0.0) && (pP.b == 0.0) && (pP.c == 0.0) {
        pOut.a = pP.a;
        pOut.b = pP.b;
        pOut.c = pP.c;
        pOut.d = pP.d;
        return Some(pOut);
    }

    // Create normalized vector (lines 13-21 in C)
    let mut n = KmVec3 {
        x: pP.a,
        y: pP.b,
        z: pP.c,
    };

    // First calculate length with immutable reference
    let length = kmVec3Length(Some(&n)).max(f32::EPSILON);
    let l = 1.0f32 / length; // Avoid division by zero
    
    // Then normalize with mutable reference
    kmVec3Normalize(Some(&mut n), None);

    pOut.a = n.x;
    pOut.b = n.y;
    pOut.c = n.z;
    pOut.d = pP.d * l;

    Some(pOut)
}
pub fn kmPlaneExtractFromMat4<'a>(pOut: Option<&'a mut KmPlane>, pIn: Option<&'a KmMat4>, row: i32) -> Option<&'a mut KmPlane> {
    // Check for None (equivalent to NULL checks in C)
    let pOut = pOut?;
    let pIn = pIn?;

    let scale = match row.cmp(&0) {
        Ordering::Less => -1.0,
        _ => 1.0,
    };
    let row = row.abs() - 1;

    // Calculate plane coefficients
    pOut.a = pIn.mat[3] + (scale * pIn.mat[row as usize]);
    pOut.b = pIn.mat[7] + (scale * pIn.mat[(row + 4) as usize]);
    pOut.c = pIn.mat[11] + (scale * pIn.mat[(row + 8) as usize]);
    pOut.d = pIn.mat[15] + (scale * pIn.mat[(row + 12) as usize]);

    // Normalize the plane and return the result
    kmPlaneNormalize(Some(pOut), None)
}
pub fn kmPlaneIntersectLine<'a>(
    pOut: Option<&'a mut KmVec3>,
    pP: Option<&'a KmPlane>,
    pV1: Option<&'a KmVec3>,
    pV2: Option<&'a KmVec3>,
) -> Option<&'a mut KmVec3> {
    // Check for None (equivalent to NULL checks in C)
    if pOut.is_none() || pP.is_none() || pV1.is_none() || pV2.is_none() {
        return pOut;
    }

    // Unwrap all Option references safely
    let pOut = pOut.unwrap();
    let pP = pP.unwrap();
    let pV1 = pV1.unwrap();
    let pV2 = pV2.unwrap();

    let mut d = KmVec3 {
        x: 0.0,
        y: 0.0,
        z: 0.0,
    };
    kmVec3Subtract(&mut d, pV2, pV1);

    let n = KmVec3 {
        x: pP.a,
        y: pP.b,
        z: pP.c,
    };
    let mut normalized_n = n.clone();
    kmVec3Normalize(Some(&mut normalized_n), Some(&n));

    let nt = -((normalized_n.x * pV1.x + normalized_n.y * pV1.y + normalized_n.z * pV1.z) + pP.d);
    let dt = normalized_n.x * d.x + normalized_n.y * d.y + normalized_n.z * d.z;

    if f32::abs(dt) < 0.0001 {
        return Some(pOut);
    }

    let t = nt / dt;
    pOut.x = pV1.x + d.x * t;
    pOut.y = pV1.y + d.y * t;
    pOut.z = pV1.z + d.z * t;

    Some(pOut)
}
pub fn kmPlaneFromPoints<'a>(pOut: &'a mut KmPlane, p1: Option<&KmVec3>, p2: Option<&KmVec3>, p3: Option<&KmVec3>) -> Option<&'a mut KmPlane> {
    // Check for NULL pointers (None in Rust)
    if p1.is_none() || p2.is_none() || p3.is_none() {
        return Option::None;
    }

    // Unwrap the options safely
    let p1 = p1.unwrap();
    let p2 = p2.unwrap();
    let p3 = p3.unwrap();

    let mut n = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut v1 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut v2 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };

    // Calculate v1 = p2 - p1
    kmVec3Subtract(&mut v1, p2, p1);
    // Calculate v2 = p3 - p1
    kmVec3Subtract(&mut v2, p3, p1);
    // Calculate n = v1 x v2 (cross product)
    kmVec3Cross(&mut n, &v1, &v2);
    
    // Create a temporary copy of n for normalization
    let n_copy = n.clone();
    // Normalize n using the copy
    kmVec3Normalize(Some(&mut n), Some(&n_copy));

    // Set plane coefficients
    pOut.a = n.x;
    pOut.b = n.y;
    pOut.c = n.z;

    // Calculate d = -n · p1
    let mut scaled_n = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    kmVec3Scale(&mut scaled_n, &n, -1.0);
    pOut.d = kmVec3Dot(Some(&scaled_n), Some(p1)).unwrap_or(0.0);

    Some(pOut)
}
pub fn kmPlaneGetIntersection<'a>(
    pOut: &'a mut KmVec3,
    p1: &KmPlane,
    p2: &KmPlane,
    p3: &KmPlane,
) -> Option<&'a mut KmVec3> {
    let mut n1 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut n2 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut n3 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut cross = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut r1 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut r2 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    let mut r3 = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
    
    kmVec3Fill(&mut n1, p1.a, p1.b, p1.c);
    kmVec3Fill(&mut n2, p2.a, p2.b, p2.c);
    kmVec3Fill(&mut n3, p3.a, p3.b, p3.c);
    
    kmVec3Cross(&mut cross, &n2, &n3);
    let denom = kmVec3Dot(Some(&n1), Some(&cross)).unwrap_or(0.0);
    
    if kmAlmostEqual(denom, 0.0) != 0 {
        return Option::None;
    }
    
    // Calculate r1, r2, r3 with temporary variables to avoid multiple mutable borrows
    let temp_r1 = {
        let mut temp = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
        kmVec3Cross(&mut temp, &n2, &n3);
        temp
    };
    kmVec3Scale(&mut r1, &temp_r1, -p1.d);
    
    let temp_r2 = {
        let mut temp = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
        kmVec3Cross(&mut temp, &n3, &n1);
        temp
    };
    kmVec3Scale(&mut r2, &temp_r2, p2.d);
    
    let temp_r3 = {
        let mut temp = KmVec3 { x: 0.0, y: 0.0, z: 0.0 };
        kmVec3Cross(&mut temp, &n1, &n2);
        temp
    };
    kmVec3Scale(&mut r3, &temp_r3, p3.d);
    
    // First subtraction
    kmVec3Subtract(pOut, &r1, &r2);
    
    // Create a temporary copy for scaling
    let mut temp = pOut.clone();
    kmVec3Scale(&mut temp, pOut, 1.0 / denom);
    
    // Final subtraction
    kmVec3Subtract(pOut, &temp, &r3);
    
    Some(pOut)
}
